platform/mbed_itm.h

```diff
@@ -0,0 +1,58 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if defined(DEVICE_ITM)
+
+#include "hal/itm_api.h"
+
+class SingleWireOutput : public FileHandle {
+public:
+    SingleWireOutput(void) {
+        itm_init();
+    }
+
+    virtual ssize_t write(const void *buffer, size_t size) {
+        const unsigned char *buf = static_cast<const unsigned char *>(buffer);
+        for (size_t i = 0; i < size; i++) {
+            ITM_SendChar(buf[i]);
+        }
+        return size;
+    }
+
+    virtual ssize_t read(void *buffer, size_t size) {
+        unsigned char *buf = static_cast<unsigned char *>(buffer);
+        buf[0] = 0;
```

Comment:
I think it's better to just return 0 (indicating "end of file") or an error code (not sure which - -EBADF = "not valid file descriptor open for reading"?; ENXIO = "... request was outside the capabilities of the device"?). Not suire which of those two choices I prefer.

My dummy device implementation inside the retarget code is only reading endless NUL bytes to preserve the legacy behaviour of stdin with DEVICE_SERIAL unset. I don't think that's actually useful, or what you should do for something new.

Reply:
Did a bit of background reading to compare other systems. This is an interesting general question.

`/dev/null` returns EOF if opened for input. But that's more likely to be used manually as a redirect for processes that expect a redirect from file, and hence understand/expect EOF on stdin.

Many real devices probably would do `ENXIO` on the attempt to open for read. I guess that means attempts to use a read-only device as process input would fail the read open and then you'd get `EBADF` later.

Linux gives `EBADF` if you read from a file that's open for output only.

When you `nohup` and detach a process in Linux it does seem you get `EBADF` on input (and you can change that to EOF if you redirect from /dev/null). POSIX says behaviour of nohup input is up to the implementation - it might leave input from original terminal, or get it from a file.

On balance, I guess I'll say `EBADF` is what to do on read. This is a "standalone" FileHandle where it's automatically opened on construction, so we should assume that the implicit open is write-only, thus give `EBADF` on read attempts.

If there was a device filing system, so we were able to participate in the open call, we'd return ENXIO on the attempt to open for read.

Reply:
Thank you! I'll change it to `EBADF`.

Reply:
Lesson learned, looks like I have some filesystems to update...

Reply:
Ah, normal filesystems might be a bit different. Eg EACCES for access violation rather than ENXIO. I was only really thinking about devices, and particularly considering this odd case of "opened by construction" objects.

Reply:
Well, in the open group docs for write:
http://pubs.opengroup.org/onlinepubs/9699919799/functions/write.html

> [EBADF]
> The fildes argument is not a valid file descriptor open for writing.

Pretty straightforward for POSIX docs actually. 😆 

A quick test program confirms this is what Linux returns.

Fortunately reading/writing to a file after opening it in a specific mode is pretty uncommon outside of user errors. I'd be worried if someone was relying on a specific error code in that situation.